src/enforcers/SwapOfferEnforcer.sol

```diff
@@ -0,0 +1,206 @@
+    }
+
+    /**
+     * @notice Decodes the terms used in this CaveatEnforcer.
+     * @param _terms encoded data that is used during the execution hooks.
+     * @return tokenIn_ The address of the token being sold.
+     * @return tokenOut_ The address of the token being bought.
+     * @return amountIn_ The total amount of tokens to be sold.
+     * @return amountOut_ The total amount of tokens to be bought.
+     * @return recipient_ The address to receive the input tokens.
+     */
+    function getTermsInfo(bytes calldata _terms) public pure returns (address tokenIn_, address tokenOut_, uint256 amountIn_, uint256 amountOut_, address recipient_) {
+        SwapOfferTerms memory terms = abi.decode(_terms, (SwapOfferTerms));
+        return (terms.tokenIn, terms.tokenOut, terms.amountIn, terms.amountOut, terms.recipient);
+    }
+
+    /**
+     * @notice Validates and updates the swap offer.
+     * @param _terms The encoded swap offer terms.
+     * @param _executionCallData The transaction the delegate might try to perform.
+     * @param _delegationHash The hash of the delegation being operated on.
+     * @param _claimedAmount The amount claimed to be transferred in.
+     * @return amountInFilled_ The updated amount of input tokens filled.
+     * @return amountOutFilled_ The updated amount of output tokens filled.
+     */
+    function _validateAndUpdate(
+        bytes calldata _terms,
+        bytes calldata _executionCallData,
+        bytes32 _delegationHash,
+        uint256 _claimedAmount
+    )
+        internal
+        returns (uint256 amountInFilled_, uint256 amountOutFilled_)
+    {
+        (address target_,, bytes calldata callData_) = _executionCallData.decodeSingle();
+
+        require(callData_.length == 68, "SwapOfferEnforcer:invalid-execution-length");
+
+        SwapOfferTerms memory terms = abi.decode(_terms, (SwapOfferTerms));
+        SwapOffer storage offer = swapOffers[msg.sender][_delegationHash];
+        if (offer.tokenIn == address(0)) {
+            // Initialize the offer if it doesn't exist
+            offer.tokenIn = terms.tokenIn;
+            offer.tokenOut = terms.tokenOut;
+            offer.amountIn = terms.amountIn;
+            offer.amountOut = terms.amountOut;
+            offer.recipient = terms.recipient;
+        } else {
+            require(offer.tokenIn == terms.tokenIn && offer.tokenOut == terms.tokenOut &&
+                    offer.amountIn == terms.amountIn && offer.amountOut == terms.amountOut &&
+                    offer.recipient == terms.recipient,
+                    "SwapOfferEnforcer:terms-mismatch");
+        }
+
+        require(target_ == terms.tokenOut, "SwapOfferEnforcer:invalid-token");
+
+        bytes4 selector = bytes4(callData_[0:4]);
+        require(selector == IERC20.transfer.selector || selector == IERC20.transferFrom.selector, "SwapOfferEnforcer:invalid-method");
+
+        uint256 amount = uint256(bytes32(callData_[36:68]));
```

Comment:
The amount should be extracted from different ranges if it is transfer [36:68] or transferFrom [68:100]